test/WebJobs.Script.Tests/WebJobsScriptHostServiceTests.cs

```diff
@@ -104,52 +82,75 @@ public async Task HostInitialization_OnInitializationException_MaintainsErrorInf
                 .Returns(hostA.Object)
                 .Returns(hostB.Object);
 
-            _hostService = new WebJobsScriptHostService(
-                _monitor, hostBuilder.Object, NullLoggerFactory.Instance, _mockRootServiceProvider.Object, _mockRootScopeFactory.Object,
-                _mockScriptWebHostEnvironment.Object, _mockEnvironment.Object, _hostPerformanceManager, _healthMonitorOptions, metricsLogger);
+            var hostService = new WebJobsScriptHostService(
+                monitor, hostBuilder.Object, NullLoggerFactory.Instance, mockRootScopeFactory.Object,
+                mockScriptWebHostEnvironment.Object, mockEnvironment.Object, hostPerformanceManager, healthMonitorOptions, metricsLogger, new Mock<IApplicationLifetime>().Object);
 
-            await _hostService.StartAsync(CancellationToken.None);
+            await hostService.StartAsync(CancellationToken.None);
             Assert.True(AreRequiredMetricsGenerated(metricsLogger));
-            Assert.Equal(ScriptHostState.Error, _hostService.State);
-            Assert.IsType<HostInitializationException>(_hostService.LastError);
+            Assert.Equal(ScriptHostState.Error, hostService.State);
+            Assert.IsType<HostInitializationException>(hostService.LastError);
         }
 
         [Fact]
         public async Task HostRestart_Specialization_Succeeds()
         {
+            var options = new ScriptApplicationHostOptions
+            {
+                ScriptPath = @"c:\tests",
+                LogPath = @"c:\tests\logs",
+            };
+            var monitor = new ScriptApplicationHostOptionsMonitor(options);
```

Comment:
why duplicating all this in each test now?

Reply:
One reason is isolation, when one test is doing a restart and some other test is doing another restart in parallel on the same instance, since we allow only one restartHostAsync in progress, the other would not happen.

---
In reply to: [369886803](https://github.com/Azure/azure-functions-host/pull/5437#discussion_r369886803) [](ancestors = 369886803)

Reply:
These tests don't run in parallel, right? I thought we disabled that for all the tests, like https://github.com/Azure/azure-functions-host/blob/7c1fb659bbce10d86e01cffcdc2b87aaeac02c15/test/WebJobs.Script.Tests.Integration/Properties/AssemblyInfo.cs#L11.

Were you seeing test failures before? I thought each test would be isolated as before

Reply:
I didn't see test failures before but that was also because RestartHostAsync() can be invoked multiple times and all of them would go through, right?. I started seeing them after my change and I thought it was because I explicitly don't allow other restarts anymore if something else is in progress already.

---
In reply to: [369892260](https://github.com/Azure/azure-functions-host/pull/5437#discussion_r369892260) [](ancestors = 369892260)

Reply:
My bad. I misunderstood, looks like the tests are fine as they are. I reverted the changes